### ISSUE TITLE
PDF/A-3u with embedfile missing /AF entry and /AFRelationship entry

### ISSUE THREAD

**Issue by wollmich** [ho-tex/oberdiek#37]
I'm having the following data file called `data.txt`:

    Testdata

and the following LaTeX code called `test2.tex`:

    \documentclass{article}
    
    \usepackage[a-3u]{pdfx}
    \usepackage{embedfile}
    \newcommand{\datafile}[2]{\embedfile[filespec={#2},ucfilespec={#2},filesystem=URL,mimetype=application/octet-stream,desc={#2},stringmethod=escape]{#1#2}}
    
    \datafile{}{data.txt}
    
    \begin{document}
    Bla bla ...
    \end{document}
    
Using `pdfLaTeX` I can generate a PDF which claims that it is PDF/A-3u.

If I verify this PDF with Adobe Preflight I get the following errors:

- Embedded file does not have /AF entry
- Embedded file has no /AFRelationship entry

[![enter image description here][1]][1]

I found the following source describing the /AF entry and the /AFRelationship entry, see https://arxiv.org/pdf/1406.6126.pdf. I do not fully understand it.

If I take a look at the uncompressed pdf file, where the embed file is described, it looks like that:

    5 0 obj
    <</Type/Filespec/FS/URL/F(data.txt)/UF(data.txt)/Desc(data.txt)/EF<</F 4 0 R>>>>
    endobj

If this would be changed to

    5 0 obj
    <</Type/Filespec/FS/URL/F(data.txt)/UF(data.txt)/Desc(data.txt)/AFRelationship/Unspecified/EF<</F 4 0 R>>>>
    endobj

then there should be no missing /AFRelationship entry anymore.

And somewhere there should be `/AF[5 0 R 5 0 R]` to solve the missing /AF entry.

Can the `embedfile` package be improved so that it adds as well the /AF entry and the /AFRelationship entry?

See as well https://tex.stackexchange.com/questions/426385/pdf-a-3u-with-embedfile-missing-af-entry-and-afrelationship-entry.

[1]: https://i.stack.imgur.com/cKbZt.png

**Comment by u-fischer** [comment]
The /AF entry is rather easy, but the /AFRelationship needs a real patch of embedfile as the current Filespec dictionary has a fix set of entries. I will think about it.

**Comment by tamersaadeh** [comment]
Are there any updated on the `/AFRelationship`? The `/AF` issue seems to have been fixed:

```latex
\documentclass{article}

% make sure it is in PDF/A format
\usepackage[a-3b]{pdfx}

% needed for pdf/a-3a conformance
%\usepackage[genericmode]{tagpdf}
%\tagpdfsetup{
%    activate-all
%}

% attached files
\usepackage{embedfile}
\embedfile[
    ucfilespec={\jobname.tex},
    filespec={\jobname.tex},
    desc={Test TeX file},
    mimetype={text/plain},
    id={test}
]{\jobname.tex}

\begin{document}

Test file.

\end{document}
```

Using [3-HEIGHTS™ PDF VALIDATOR](https://www.pdf-online.com/osa/validate.aspx), the issues are:

|File | test.pdf|
-- | --
Compliance | pdfa-3b
Result | Document does not conform to PDF/A.
Details | Validating file "test.pdf" for conformance level pdfa-3b
||The key AFRelationship is required but missing.
||File specification 'test.tex' not associated with an object.
||The document does not conform to the requested standard.
||The document doesn't conform to the PDF reference (missing required entries, wrong value types, etc.).
||The document does not conform to the PDF/A-3b standard.Done.


The same result is obtained when testing with the more strict PDF/A-3A specification.

**Comment by u-fischer** [comment]
@tamersaadeh I'm planing to rewrite embedfile (I need the keys too in the near future), but I can't say when I will find the time.

**Comment by akarrenbauer** [comment]
I stumbled over this issue when looking for a solution to create [ZUGFeRD](https://www.ferd-net.de/) compliant hybrid invoices using pdflatex. To this end, an xml-file has to be embedded in a PDF/A-3 compliante file.

As in the examples above, I use the pdfx package.

After some trial&error hacking embedfile.sty, I could make my PDFs to pass validation using veraPDF.

I've created a pull request with my patch. I have to admit that I'm using Latex for decades, but that this was the first time that I've digged into a sty-file. Hence, my patch probably won't pass quality control. However, it solved my issues and could potentially serve as a proof-of-concept for a cleaner solution.

The previous example above can be made to pass validation for PDF/A-3b compliance as follows:
```
\documentclass{article}

% make sure it is in PDF/A format
\usepackage[a-3b]{pdfx}

% needed for pdf/a-3a conformance
%\usepackage[genericmode]{tagpdf}
%\tagpdfsetup{
%    activate-all
%}

% attached files
\usepackage{embedfile}
\embedfile[
    ucfilespec={\jobname.tex},
    filespec={\jobname.tex},
    desc={Test TeX file},
    mimetype={text/plain},
    id={test},
    afrelationship={/Source}
]{\jobname.tex}

\begin{document}

Test file.

\end{document}
```

**Comment by UncertaintyP** [comment]
Exactly what I need, tyvm @akarrenbauer 

One note for ZUGFeRD - the provided example files use `/Alternative` instead of `/Source` for `afrelationship` but both are compliant :)

Edit: According to the standard they must use `/Alternative`

**Comment by tecosaur** [comment]
Is anything likely to happen with this? It seems like this issue is just sitting here.

**Comment by davidcarlisle** [comment]
Something is likely to happen but you are asking people to use their spare time to update a suite of packages that were adopted when the original author stopped maintaining them, and just running the tests and repackaging a ctan upload takes well over an hour, even if there are no changes, so finding a slot where there is time to look at this means you can not necessarily expect a response in a few days to any issue posted.

One of the main reasons why the packages were moved to a public tracker is that any patches or problems seen can be used by others even if they are not yet incorporated back to the main release.

**Comment by u-fischer** [comment]
@tecosaur  Yes, there will happen something. But the main work is currently done in one of the branches of the pdfresource project on the latex3 site, it implements a number of commands and interface that I will need before I can tackle the task to change/reimplement embedfile (and perhaps merge it with attachfile2). In the mean time you can try the code in the pull request. It looks ok.

**Comment by tecosaur** [comment]
I see, that sounds entirely reasonable. Would editing the `.sty` on my machine be a bad idea?
